src/pointcloud_to_laserscan_nodelet.cpp

```diff
@@ -168,74 +238,145 @@ namespace pointcloud_to_laserscan
+   */
+  void scan_outlier_removal_filter(sensor_msgs::LaserScan &scan, double cluster_break_distance, int max_noise_cluster_size)
+  {
+      // help function initialization
+      int ranges_size = scan.ranges.size();
+      int i_current_cluster = 0;
+      bool cluster_further_away = false;
+      std::vector<int> cluster_indecies;
+      double diffs[ranges_size-1];
 
-      if (*iter_z > max_height_ || *iter_z < min_height_)
+      // calculate diffs
+      for (int i = 1; i< ranges_size; i++)
       {
-        NODELET_DEBUG("rejected for height %f not in range (%f, %f)\n", *iter_z, min_height_, max_height_);
-        continue;
+          diffs[i-1] = scan.ranges[i] - scan.ranges[i-1];
       }
 
-      double range = hypot(*iter_x, *iter_y);
-      if (range < range_min_)
+      // find and remove outliers
+      for (int i = 0; i< ranges_size-1; i++)
       {
-        NODELET_DEBUG("rejected for range %f below minimum value %f. Point: (%f, %f, %f)", range, range_min_, *iter_x, *iter_y,
-                      *iter_z);
-        continue;
-      }
+          // add point to cluster
+          cluster_indecies.push_back(i);
+          i_current_cluster ++;
 
-      double angle = atan2(*iter_y, *iter_x);
-      if (angle < output.angle_min || angle > output.angle_max)
-      {
-        NODELET_DEBUG("rejected for angle %f not in range (%f, %f)\n", angle, output.angle_min, output.angle_max);
-        continue;
+          // check if last point in cluster; find diff larger than border -> cluster separation
+          if (diffs[i] > cluster_break_distance || diffs[i] < -cluster_break_distance)
+          {
+              bool new_cluster_further_away = (diffs[i] > 0);
+
+              // Only remove cluster if it is closer than surrounding clusters
+              if ((i_current_cluster < max_noise_cluster_size) && (new_cluster_further_away != cluster_further_away))
+              {
+                  for (int k = 0; k < i_current_cluster; k++)
+                  {
+                      scan.ranges[cluster_indecies[k]] = scan.range_max;
+                  }
+              }
+
+              // initialize new cluster
+              cluster_further_away = new_cluster_further_away;
+              cluster_indecies.clear();
+              i_current_cluster = 0;
+          }
       }
 
-      //overwrite range at laserscan ray if new range is smaller
-      int index = (angle - output.angle_min) / output.angle_increment;
-      if (range < output.ranges[index])
+
+      // remove last cluster if small enough
+      if ((i_current_cluster < cluster_break_distance) && !cluster_further_away)
       {
-        output.ranges[index] = range;
+          for (int k = 0; k < i_current_cluster; k++)
+          {
+              scan.ranges[cluster_indecies[k]] = scan.range_max;
+          }
+          // remove last point as well
+          scan.ranges[ranges_size-1] = scan.range_max;
       }
-
-    }
-    pub_.publish(output);
   }
-
```

Comment:
the changes in this file actually look like a complete re-write....I do not understand the changes/functionality, but it is quite difficult to review...if it is a re-write and therewith changing the overall functionallity (that other might rely on), we might want to add "our" new nodelet as an additional node?